// wired-lib/lib/wired-base.js
export { TemplateResult } from '../../lit-html/lib/template-result.js';
export { html } from '../../lit-html/lit-html.js';
export { UpdatingElement, defaultConverter, notEqual } from '../../lit-element/lib/updating-element.js';
export { customElement, property, query } from '../../lit-element/lib/decorators.js';
export { CSSResult, css, supportsAdoptingStyleSheets } from '../../lit-element/lib/css-tag.js';
import { LitElement } from '../../lit-element/lit-element.js';
export { LitElement } from '../../lit-element/lit-element.js';

class WiredBase extends LitElement {
    fireEvent(name, detail, bubbles = true, composed = true) {
        if (name) {
            const init = {
                bubbles: (typeof bubbles === 'boolean') ? bubbles : true,
                composed: (typeof composed === 'boolean') ? composed : true
            };
            if (detail) {
                init.detail = detail;
            }
            const CE = (window.SlickCustomEvent || CustomEvent);
            this.dispatchEvent(new CE(name, init));
        }
    }
}

export { WiredBase };
